### PR TITLE
Add MPI support for Pytorch distributed training

### PR DESCRIPTION
**Pull request by hkvision**
@jason-dai Take a look at the initial implementation.

__MPIRunner__ is a general class for running any mpi script. (Probably we can support some optimized configurations in the later stage, refer to https://github.com/intel/intel-extension-for-pytorch/blob/master/intel_pytorch_extension_py/launch.py

__MPIEstimator__ is an Estimator for Pytorch distributed training with MPI. (No extra Python script is needed in this case) Fit accepts Spark DataFrames and put into plasma, and would create a dataloader based on numpy for users to iterate. Users have the flexibility to define init_func, train_func and validate_func, if not specified there are default ones. Users can also input data_creator if they wish.

Verified on dlrm e2e workload on clx cluster and 4-socket cpx.

### CODE REVIEW

**Comment by hkvision** [comment]
The current settings is for oneCCL.
For OpenMPI, may need to change the command `mpiexec.hydra` to `mpiexec.openmpi` and the environment variable to `PMIX_RANK`

**Comment by hkvision** [comment]
@jason-dai This PR has been almost finished. Take a look. I'm going to first check-in an initial version.

**Comment by hkvision** [comment]
https://github.com/analytics-zoo/friesian/pull/31 See the dlrm example here

**Comment by hkvision** [comment]
jenkins passed: http://10.239.47.210:18888/job/ZOO-PR-Validation/5306/

File: pyzoo/zoo/orca/learn/mpi/mpi_runner.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+            def get_ip(iter):
+                yield get_node_ip()
+
+            from zoo.orca import OrcaContext
+            sc = OrcaContext.get_spark_context()
+            master = sc.getConf().get("spark.master")
+            if master == "local" or master.startswith("local["):
+                num_executors = 1
+            else:
+                num_executors = int(sc.getConf().get("spark.executor.instances"))
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/mpi/mpi_runner.py:46]: use `get_node_and_core_number`?

File: pyzoo/zoo/orca/learn/mpi/mpi_runner.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            self.hosts = list(set(sc.range(0, num_executors, numSlices=num_executors).barrier()
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/mpi/mpi_runner.py:47]: I don't think this is correct on a cluster; there is no guarantee that you will run a task on each node

File: pyzoo/zoo/orca/learn/mpi/mpi_runner.py (hunk 1 above)
Comment (hkvision) [reply pyzoo/zoo/orca/learn/mpi/mpi_runner.py:46]: okay

File: pyzoo/zoo/orca/learn/mpi/mpi_runner.py (hunk 2 above)
Comment (hkvision) [reply pyzoo/zoo/orca/learn/mpi/mpi_runner.py:47]: oh really? we use this code to launch ray alongside each executor?

Comment (hkvision) [reply pyzoo/zoo/orca/learn/mpi/mpi_runner.py:47]: https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/ray/raycontext.py#L504

File: pyzoo/zoo/orca/learn/mpi/utils.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    else:
+        if end:
+            return x[:end]
+        else:
+            return x
+
+
+def process_records(buffer):
+    import random
+    random.shuffle(buffer)  # TODO: Make shuffle configurable?
```
Comment (hkvision) [inline pyzoo/zoo/orca/learn/mpi/utils.py:48]: shuffle is here @jason-dai

**Comment by hkvision** [comment]
cloudpickle 1.5.0; pyarrow 0.17.0
